datacube/testutils/__init__.py

```diff
@@ -237,6 +247,8 @@ def with_keys(d, keys):
                                 measurements=measurements,
                                 **product_opts)
 
+    kwargs: dict[str, str | None]
```

Comment:
This signature is added a few lines down on the first assignment in #1785.

I also think this annotation is missing the list type in the dictionary.